### ISSUE TITLE
Install fails when using `x-scripts-root`

### ISSUE THREAD

**Issue by vicroms** [microsoft/vcpkg#8160]
**Describe the bug**
When using `--x-scripts-root` packages fail to configure if no `.vcpkg-root` file exists in a directory above the specified scripts root.

The failure happens in `scripts/buildsystems/vcpkg.cmake`.

**Environment**
- OS: All
- Compiler: N/A

**To Reproduce**

Starting from vcpkg repo
1. cd ..
2. cp -r ./vcpkg/scripts ./scripts
3. cd vcpkg
4. ./vcpkg install xyz --scripts-root=../scripts

Build will fail with a message about `.vcpkg-root` not being found.

**Expected behavior**
A successful package installation.

**Additional context**
Path to vcpkg root is known by the vcpkg executable, this path could be passed to the build in order to avoid this failure.

**Comment by martin-s** [comment]
Ok, did not have much time last week. I think the forwarding of vcpkg-root path is not working as planned. Before, lookup started from triplet folder, too.

**Comment by martin-s** [comment]
I can only think of using variable _CMAKE_INSTALL_PREFIX_ on a second lookup. This is pointing to the package folder of the currently build port.
Otherwise, variable _VCPKG_ROOT_PATH_ could be forwarded as proposed which means introducing another variable in the cmake toolchain workflow.
It looks like it's used only in file _vcpkg_configure_cmake.cmake_. Hopefully, not in other contexts.

**Comment by Cheney-W** [comment]
Verified on the latest VCPKG 2021-07-26-9425cf5f512f242c0bcbabac31f08832825aee81 , the installation step will hung instead of fail with a message about .vcpkg-root not being found.

**Comment by invy** [comment]
TL;DR
Please don't remove --x-scripts-root and return the help back as it is a useful option and it works.


As of now I can't actually confirm the bug.
@vicroms could you reproduce the bug with the current version of vcpkg-tool (from git)?

I suppose the bug was fixed by some other commits.

Consider following output/scenario:

vcpkg_test_root is an empty dir.
vcpkgData contains triplets, scripts and ports.
Starting here:

```
vcpkg_test_root$ ./vcpkg install zlib --overlay-ports=../vcpkgData/ports --overlay-triplets=../vcpkgData/triplets --x-scripts-root=../vcpkgData/scripts
Computing installation plan...
A suitable version of cmake was not found (required v3.24.0). Downloading portable cmake v3.24.0...
Downloading cmake...
  https://github.com/Kitware/CMake/releases/download/v3.24.0/cmake-3.24.0-linux-x86_64.tar.gz -> /home/igor/development/vcpkg_test_root/downloads/cmake-3.24.0-linux-x86_64.tar.gz
Extracting cmake...
The following packages will be built and installed:
    zlib[core]:x64-linux -> 1.2.11-5 -- /home/igor/development/vcpkg_test_root/../vcpkgData/ports/zlib
Detecting compiler hash for triplet x64-linux...
Restored 0 package(s) from /home/igor/.cache/vcpkg/archives in 5.188 us. Use --debug to see more details.
Installing 1/1 zlib:x64-linux...
Building zlib[core]:x64-linux...
-- [OVERLAY] Loading triplet configuration from: /home/igor/development/vcpkgData/triplets/x64-linux.cmake
-- Installing port from location: /home/igor/development/vcpkg_test_root/../vcpkgData/ports/zlib
CMake Warning at /home/igor/development/vcpkgData/scripts/cmake/vcpkg_common_functions.cmake:3 (message):
  vcpkg_common_functions has been removed and all values are automatically
  provided in all portfile.cmake invocations.  Please remove
  `include(vcpkg_common_functions)`.
Call Stack (most recent call first):
  /home/igor/development/vcpkgData/ports/zlib/portfile.cmake:1 (include)
  /home/igor/development/vcpkgData/scripts/ports.cmake:147 (include)


-- Downloading http://www.zlib.net/zlib-1.2.11.tar.gz;https://downloads.sourceforge.net/project/libpng/zlib/1.2.11/zlib-1.2.11.tar.gz -> zlib1211.tar.gz...
-- Extracting source /home/igor/development/vcpkg_test_root/downloads/zlib1211.tar.gz
-- Applying patch cmake_dont_build_more_than_needed.patch
-- Using source at /home/igor/development/vcpkg_test_root/buildtrees/zlib/src/1.2.11-f690224aeb.clean
-- Configuring x64-linux-dbg
-- Configuring x64-linux-rel
-- Building x64-linux-dbg
-- Building x64-linux-rel
-- Fixing pkgconfig file: /home/igor/development/vcpkg_test_root/packages/zlib_x64-linux/lib/pkgconfig/zlib.pc
-- Fixing pkgconfig file: /home/igor/development/vcpkg_test_root/packages/zlib_x64-linux/debug/lib/pkgconfig/zlib.pc
-- Installing: /home/igor/development/vcpkg_test_root/packages/zlib_x64-linux/share/zlib/copyright
CMake Warning at /home/igor/development/vcpkgData/scripts/cmake/vcpkg_test_cmake.cmake:3 (message):
  vcpkg_test_cmake was a no-op and has been removed.  Please remove the call
  to `vcpkg_test_cmake()`.
Call Stack (most recent call first):
  /home/igor/development/vcpkgData/ports/zlib/portfile.cmake:70 (vcpkg_test_cmake)
  /home/igor/development/vcpkgData/scripts/ports.cmake:147 (include)


-- Performing post-build validation
-- Performing post-build validation done
Stored binary cache: "/home/igor/.cache/vcpkg/archives/b3/b3c649f694ac3c5507726abd6fe5d4f0cd55041228387ef133d1b0fc79669e6a.zip"
Elapsed time to handle zlib:x64-linux: 29.08 s

Total elapsed time: 36.89 s

The package zlib is compatible with built-in CMake targets:

    find_package(ZLIB REQUIRED)
    target_link_libraries(main PRIVATE ZLIB::ZLIB)


vcpkg_test_root$ ls -a
.  ..  buildtrees  downloads  installed  packages  vcpkg  .vcpkg-root
vcpkg_test_root$ ls -a ../vcpkgData/
.  ..  .git  ports  scripts  triplets
```
So I don't see the bug here actually. If you are overriding the parameters you should be aware of what you are doing.
Considering all this I don't see the reason to remove --x-scripts-root.

**Comment by invy** [comment]
Especially please leave the parameter because of the symmetry and because you can do a cool stuff with it:
```
vcpkg_test_root$ mkdir x-root
vcpkg_test_root$ ./vcpkg install zlib --overlay-ports=../vcpkgData/ports --overlay-triplets=../vcpkgData/triplets --x-scripts-root=../vcpkgData/scripts --vcpkg-root=./x-root --x-packages-root=./x-pk --downloads-root=./x-dl --x-install-root=./x-inst --x-buildtrees-root=./x-bs
Computing installation plan...
A suitable version of cmake was not found (required v3.24.0). Downloading portable cmake v3.24.0...
Downloading cmake...
  https://github.com/Kitware/CMake/releases/download/v3.24.0/cmake-3.24.0-linux-x86_64.tar.gz -> /home/igor/development/vcpkg_test_root/x-dl/cmake-3.24.0-linux-x86_64.tar.gz
Extracting cmake...
The following packages will be built and installed:
    zlib[core]:x64-linux -> 1.2.11-5 -- /home/igor/development/vcpkg_test_root/../vcpkgData/ports/zlib
Detecting compiler hash for triplet x64-linux...
Restored 0 package(s) from /home/igor/.cache/vcpkg/archives in 3.349 us. Use --debug to see more details.
Installing 1/1 zlib:x64-linux...
Building zlib[core]:x64-linux...
-- [OVERLAY] Loading triplet configuration from: /home/igor/development/vcpkgData/triplets/x64-linux.cmake
-- Installing port from location: /home/igor/development/vcpkg_test_root/../vcpkgData/ports/zlib
CMake Warning at /home/igor/development/vcpkgData/scripts/cmake/vcpkg_common_functions.cmake:3 (message):
  vcpkg_common_functions has been removed and all values are automatically
  provided in all portfile.cmake invocations.  Please remove
  `include(vcpkg_common_functions)`.
Call Stack (most recent call first):
  /home/igor/development/vcpkgData/ports/zlib/portfile.cmake:1 (include)
  /home/igor/development/vcpkgData/scripts/ports.cmake:147 (include)


-- Downloading http://www.zlib.net/zlib-1.2.11.tar.gz;https://downloads.sourceforge.net/project/libpng/zlib/1.2.11/zlib-1.2.11.tar.gz -> zlib1211.tar.gz...
-- Extracting source /home/igor/development/vcpkg_test_root/x-dl/zlib1211.tar.gz
-- Applying patch cmake_dont_build_more_than_needed.patch
-- Using source at /home/igor/development/vcpkg_test_root/x-bs/zlib/src/1.2.11-f690224aeb.clean
-- Configuring x64-linux-dbg
-- Configuring x64-linux-rel
-- Building x64-linux-dbg
-- Building x64-linux-rel
-- Fixing pkgconfig file: /home/igor/development/vcpkg_test_root/x-pk/zlib_x64-linux/lib/pkgconfig/zlib.pc
-- Fixing pkgconfig file: /home/igor/development/vcpkg_test_root/x-pk/zlib_x64-linux/debug/lib/pkgconfig/zlib.pc
-- Installing: /home/igor/development/vcpkg_test_root/x-pk/zlib_x64-linux/share/zlib/copyright
CMake Warning at /home/igor/development/vcpkgData/scripts/cmake/vcpkg_test_cmake.cmake:3 (message):
  vcpkg_test_cmake was a no-op and has been removed.  Please remove the call
  to `vcpkg_test_cmake()`.
Call Stack (most recent call first):
  /home/igor/development/vcpkgData/ports/zlib/portfile.cmake:70 (vcpkg_test_cmake)
  /home/igor/development/vcpkgData/scripts/ports.cmake:147 (include)


-- Performing post-build validation
-- Performing post-build validation done
Stored binary cache: "/home/igor/.cache/vcpkg/archives/b3/b3c649f694ac3c5507726abd6fe5d4f0cd55041228387ef133d1b0fc79669e6a.zip"
Elapsed time to handle zlib:x64-linux: 3.496 s

Total elapsed time: 10.18 s

The package zlib is compatible with built-in CMake targets:

    find_package(ZLIB REQUIRED)
    target_link_libraries(main PRIVATE ZLIB::ZLIB)

```

**Comment by BillyONeal** [comment]
> TL;DR
> Please don't remove --x-scripts-root and return the help back as it is a useful option and it works.

The reason we are likely to remove `--x-scripts-root` is that everything in the scripts root used by vcpkg is 'morally' part of the tool, and its presence in our default curated registry a historical artifact of before the curated registry and the tool were in separate repos.

>So I don't see the bug here actually. If you are overriding the parameters you should be aware of what you are doing.
>Considering all this I don't see the reason to remove --x-scripts-root.

I'm not seeing anything here that wouldn't be better served by leaving the VCPKG_ROOT/scripts relationship alone and changing the installed tree root instead. To be clear, installed root, packages root, and downloads root are all reasonable things for users to customize and I think we should drop the x- from them. But basically nothing can be done with x-scripts-root that we can meaningfully support. coordinated changes between the tool binary and the scripts are fair game, which means you must use exactly the same scripts as match that vcpkg binary, which means there is very little argument to override it to something else.

>Especially please leave the parameter because of the symmetry and because you can do a cool stuff with it.
>This opens plenty of possibilities, when you building packages on CI in own environment.
>It also allows to keep repositories clean and have the control over where data are being extracted.

You changed vcpkg-root in this example so it would still be out of the source tree or similar.

**Comment by invy** [comment]
@BillyONeal 
> The reason we are likely to remove --x-scripts-root is that everything in the scripts root used by vcpkg is 'morally' part of the tool

Yep, that I can totally agree on. Basically in my use case I'm apparently to keep for each release Tool and corresponding scripts (or rather vcpkg branch where do scripts reside) to be able to build my 3rd party packages from scratch.

> I'm not seeing anything here that wouldn't be better served by leaving the VCPKG_ROOT
> You changed vcpkg-root in this example so it would still be out of the source tree or similar.

Well, you're right. If I'm overriding almost everything anyway including managing my overlay ports and triplets.

If I just move scripts to the repository where my overlay ports reside and call it my new vcpkg root. I could additionally modify booststrap and get everytime a  fitting version of vcpkg tool for the residing in the corresponding branch.
It would be actually even better solution then just overriding scripts ports and triplets.

**Comment by senstar-nross** [comment]
I just wanted to add to the mention above https://github.com/microsoft/vcpkg/issues/22074#issuecomment-2148135685. 

> > TL;DR
> > Please don't remove --x-scripts-root and return the help back as it is a useful option and it works.
> 
> The reason we are likely to remove `--x-scripts-root` is that everything in the scripts root used by vcpkg is 'morally' part of the tool, and its presence in our default curated registry a historical artifact of before the curated registry and the tool were in separate repos.
> 
> > So I don't see the bug here actually. If you are overriding the parameters you should be aware of what you are doing.
> > Considering all this I don't see the reason to remove --x-scripts-root.
> 
> I'm not seeing anything here that wouldn't be better served by leaving the VCPKG_ROOT/scripts relationship alone and changing the installed tree root instead. To be clear, installed root, packages root, and downloads root are all reasonable things for users to customize and I think we should drop the x- from them. But basically nothing can be done with x-scripts-root that we can meaningfully support. coordinated changes between the tool binary and the scripts are fair game, which means you must use exactly the same scripts as match that vcpkg binary, which means there is very little argument to override it to something else.
> 

@BillyONeal I do understand that the scripts morally belong to the tool, but IMO the scripts immorally impact the ABI and therefore make binary caching unstable.  

I fail to see in most cases how the powershell version impacts the ABI of the cached package. And honestly in most cases the cmake version etc also has very little impact too.  

If I build an opencv4, 2 year old version and cache it and then rebuild the exact same version with a new vcpkg version I understand that it *could* be different but for me and my team I don't care I just want to keep using the version I have cached.  

And morally it should be the same otherwise vcpkg is breaking things in very unexpected ways and would quickly be deemed an unreliable tool... If I use the same compiler, the same msbuild, make, cmake files etc.. to compile a 2 year old version of source code I expect it to be the *same*.  

I mean if we get right down to it, an update to your OS is probably as likely to affect the ABI as the powershell version.  The clock is different, the timestamps in the libraries are different, some header file has __TIME__ in it...
